codec-redis/src/main/java/io/netty/handler/codec/redis/RedisDecoder.java

```diff
@@ -0,0 +1,232 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.netty.handler.codec.redis;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.codec.CodecException;
+import io.netty.util.ByteProcessor;
+
+import java.util.List;
+
+/**
+ * Decodes the Redis protocol into {@link RedisMessage} objects following
+ * <a href="http://redis.io/topics/protocol">RESP (REdis Serialization Protocol)</a>.
+ *
+ * {@link RedisMessage} parts can be aggregated to {@link RedisMessage} using
+ * {@link RedisMessageAggregator} or processed directly.
+ */
+public class RedisDecoder extends ByteToMessageDecoder {
+
+    private static final int CRLF_LENGTH = 2;
+    private static final long NULL_RESPONSE = -1;
+
+    private final ToLongProcessor toLongProcessor = new ToLongProcessor();
+
+    // current decoding states
+    private State state;
+    private RedisMessageType type;
+
+    // for decoding bulk string
+    private long bulkStringLength;
+
+    private enum State {
+        DECODE_TYPE,
+        DECODE_INLINE, // SIMPLE_STRING, ERROR, INTEGER
+        DECODE_LENGTH, // BULK_STRING, ARRAY_HEADER
+        DECODE_BULK_STRING,
+    }
+
+    public RedisDecoder() {
+        resetDecoder();
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        try {
+            boolean next;
+            do {
+                switch (state) {
+                case DECODE_TYPE:
+                    next = decodeType(in);
+                    break;
+                case DECODE_INLINE:
+                    next = decodeInline(in, out);
+                    break;
+                case DECODE_LENGTH:
+                    next = decodeLength(in, out);
+                    break;
+                case DECODE_BULK_STRING:
+                    next = decodeBulkString(in, out);
+                    break;
+                default:
+                    throw new CodecException("Unknown state: " + state);
+                }
+            } while (next);
+        } catch (Exception e) {
+            resetDecoder();
+            throw new CodecException(e);
+        }
+    }
+
+    private void resetDecoder() {
+        state = State.DECODE_TYPE;
+    }
+
+    private boolean decodeType(ByteBuf in) throws Exception {
+        if (in.readableBytes() < 1) {
+            return false;
+        }
+        type = RedisMessageType.valueOf(in.readByte());
+        state = type.isInline() ? State.DECODE_INLINE : State.DECODE_LENGTH;
+        return true;
+    }
+
+    private boolean decodeInline(ByteBuf in, List<Object> out) throws Exception {
+        ByteBuf lineBytes = readLine(in);
+        if (lineBytes == null) {
+            return false;
```

Comment:
if we can not find the EOF, and the content is bigger than a threshold (`512mb` by default but configurable in the constructor?) we should consider throwing/closing the connection.